eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Upec.hs

```diff
@@ -144,7 +146,7 @@ votedValue (ProposedPPUpdates pup) pps quorumN =
         --   1) `consensus` is empty, or
         --   2) `consensus` has exactly one element.
         1 ->
-          (Just . applyPPUpdates pps . fst . head . Map.toList)
+          (Just . applyPPUpdates pps . fst . fst . fromJust . List.uncons . Map.toList)
```

Comment:
Original implementation is pretty bad.  But the suggested one is not better either. Could you please switch back to `head` so we don't forget to fix it properly.
```suggestion
          (Just . applyPPUpdates pps . fst . head . Map.toList)
```

Proper fix is this:
```haskell
   in case Map.keys consensus of
        -- NOTE that `quorumN` is a global constant, and that we require
        -- it to be strictly greater than half the number of genesis nodes.
        -- The keys in the `pup` correspond to the genesis nodes,
        -- and therefore either:
        --   1) `consensus` is empty, or
        --   2) `consensus` has exactly one element.
        thePPups:_ -> Just $ applyPPUpdates pps thePPups
        -- NOTE that `updatePParams` corresponds to the union override right
        -- operation in the formal spec.
        _ -> Nothing
```

Reply:
Reverted to the original code.